File: agenkit-ts/src/middleware/retry.ts
/**
 * Retry middleware - automatic retries with exponential backoff.
 *
 * Handles transient failures by retrying requests with configurable
 * backoff strategy.
 */

import { Agent, Message } from '../core/interfaces';
import { BaseMiddleware } from './base';

/**
 * Retry configuration.
 */
export interface RetryConfig {
  /** Maximum number of retry attempts (default: 3) */
  maxAttempts?: number;

  /** Initial delay in milliseconds (default: 1000) */
  initialDelay?: number;

  /** Backoff multiplier (default: 2.0) */
  backoffMultiplier?: number;

  /** Maximum delay in milliseconds (default: 30000) */
  maxDelay?: number;

  /** Predicate to determine if error should trigger retry */
  shouldRetry?: (error: Error) => boolean;
}

/**
 * Retry metrics.
 */
export interface RetryMetrics {
  /** Total number of requests (including retries) */
  totalAttempts: number;

  /** Number of requests that succeeded on first try */
  successfulFirstAttempt: number;

  /** Number of requests that succeeded after retry */
  successfulOnRetry: number;

  /** Number of requests that failed after all retries */
  failedAfterRetries: number;

  /** Total number of retry attempts across all requests */
  totalRetries: number;
}

/**
 * Default retry predicate - retries on network errors.
 */
function defaultShouldRetry(error: Error): boolean {
  // Retry on network errors, timeouts, 5xx errors
  const message = error.message.toLowerCase();
  return (
    message.includes('network') ||
    message.includes('timeout') ||
    message.includes('econnrefused') ||
    message.includes('enotfound') ||
    message.includes('http 5')
  );
}

/**
 * RetryMiddleware implements automatic retry with exponential backoff.
 *
 * Features:
 * - Configurable retry attempts
 * - Exponential backoff
 * - Customizable retry predicate
 * - Preserves error stack traces
 *
 * Usage:
 *   const agent = new RetryMiddleware(baseAgent, {
 *     maxAttempts: 3,
 *     initialDelay: 1000,
 *     backoffMultiplier: 2.0,
 *   });
 */
export class RetryMiddleware extends BaseMiddleware {
  private maxAttempts: number;
  private initialDelay: number;
  private backoffMultiplier: number;
  private maxDelay: number;
  private shouldRetry: (error: Error) => boolean;
  private _metrics: RetryMetrics;

  constructor(agent: Agent, config: RetryConfig = {}) {
    super(agent);
    this.maxAttempts = config.maxAttempts || 3;
    this.initialDelay = config.initialDelay || 1000;
    this.backoffMultiplier = config.backoffMultiplier || 2.0;
    this.maxDelay = config.maxDelay || 30000;
    this.shouldRetry = config.shouldRetry || defaultShouldRetry;
    this._metrics = {
      totalAttempts: 0,
      successfulFirstAttempt: 0,
      successfulOnRetry: 0,
      failedAfterRetries: 0,
      totalRetries: 0,
    };
  }

  /**
   * Get current metrics.
   */
  get metrics(): RetryMetrics {
    return { ...this._metrics };
  }

  async process(message: Message): Promise<Message> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.maxAttempts; attempt++) {
      this._metrics.totalAttempts++; // Track each attempt (initial + retries)

      try {
        const result = await this.agent.process(message);

        // Record success
        if (attempt === 0) {
          this._metrics.successfulFirstAttempt++;
        } else {
          this._metrics.successfulOnRetry++;
        }

        return result;
      } catch (error) {
        lastError = error as Error;

        // Track retry attempts (not counting the initial attempt)
        if (attempt > 0) {
          this._metrics.totalRetries++;
        }

        // Don't retry if predicate says no or if we're on last attempt
        if (!this.shouldRetry(lastError) || attempt === this.maxAttempts - 1) {
          this._metrics.failedAfterRetries++;
          throw lastError;
        }

        // Calculate delay with exponential backoff
        const delay = Math.min(
          this.initialDelay * Math.pow(this.backoffMultiplier, attempt),
          this.maxDelay,
        );

        // Wait before retrying
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }

    // Should never reach here, but TypeScript needs it
    this._metrics.failedAfterRetries++;
    throw lastError || new Error('Retry failed');
  }
}

/**
 * Create retry middleware function.
 *
 * @param config Retry configuration
 * @returns Middleware function
 */
export function retry(config: RetryConfig = {}) {
  return (agent: Agent): Agent => new RetryMiddleware(agent, config);
}

File: agenkit-cpp/src/middleware/retry.cpp
/**
 * @file retry.cpp
 * @brief Retry middleware implementation
 */

#include "agenkit/middleware/retry.hpp"
#include <cmath>
#include <algorithm>

namespace agenkit {
namespace middleware {

std::chrono::milliseconds RetryMiddleware::calculate_backoff(uint32_t attempt) {
    // Calculate exponential backoff: initial * (multiplier ^ attempt)
    double backoff_ms = config_.initial_backoff.count() *
                        std::pow(config_.backoff_multiplier, static_cast<double>(attempt));

    // Clamp to max_backoff
    backoff_ms = std::min(backoff_ms, static_cast<double>(config_.max_backoff.count()));

    // Add jitter if enabled (0-50% of the calculated backoff)
    if (config_.enable_jitter) {
        std::uniform_real_distribution<double> dist(0.0, 0.5);
        double jitter_factor = dist(rng_);
        backoff_ms = backoff_ms * (1.0 + jitter_factor);
    }

    return std::chrono::milliseconds(static_cast<int64_t>(backoff_ms));
}

std::future<core::Result<core::Message, core::AgentError>>
RetryMiddleware::process(core::Message message) {
    metrics_.total_attempts++;

    bool had_retry = false;
    core::Result<core::Message, core::AgentError> last_result =
        core::Result<core::Message, core::AgentError>::err(
            core::AgentError(core::AgentErrorType::Internal, "No attempts made")
        );

    for (uint32_t attempt = 0; attempt < config_.max_attempts; ++attempt) {
        // Try the operation
        auto future = agent_->process(message);
        auto result = future.get();

        // Success - return immediately
        if (result.is_ok()) {
            if (had_retry) {
                metrics_.successful_on_retry++;
            }
            return core::make_ready_future(std::move(result));
        }

        // Got error - save for potential return
        last_result = std::move(result);

        // Don't retry if we're on the last attempt
        if (attempt + 1 >= config_.max_attempts) {
            break;
        }

        // Check retry predicate
        if (!config_.should_retry(last_result.unwrap_err())) {
            break;
        }

        // We're retrying
        had_retry = true;
        metrics_.total_retries++;

        // Calculate and apply backoff
        auto backoff = calculate_backoff(attempt);
        metrics_.total_backoff_ms += backoff.count();

        std::this_thread::sleep_for(backoff);
    }

    // All retries exhausted
    metrics_.failed_after_retries++;
    return core::make_ready_future(std::move(last_result));
}

} // namespace middleware
} // namespace agenkit
